src/main/java/com/aws/greengrass/lifecyclemanager/UnloadableService.java

```diff
@@ -0,0 +1,103 @@
+import com.aws.greengrass.util.Coerce;
+import com.aws.greengrass.util.NucleusPaths;
+import com.aws.greengrass.util.Utils;
+import com.vdurmont.semver4j.Semver;
+import com.vdurmont.semver4j.SemverException;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.time.Instant;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+
+import static com.amazon.aws.iot.greengrass.component.common.ComponentType.PLUGIN;
+import static com.aws.greengrass.componentmanager.KernelConfigResolver.VERSION_CONFIG_KEY;
+import static com.aws.greengrass.dependency.EZPlugins.JAR_FILE_EXTENSION;
+import static com.aws.greengrass.deployment.bootstrap.BootstrapSuccessCode.REQUEST_RESTART;
+import static com.aws.greengrass.lifecyclemanager.Kernel.SERVICE_TYPE_TOPIC_KEY;
+
+public class UnloadableService extends GreengrassService {
+    ServiceLoadException serviceLoadException;
+    long exceptionTimestamp;
+
+    /**
+     * Create a new instance for unloadable service.
+     *
+     * @param topics root topic for this service
+     * @param e the original exception while loading the service
+     */
+    public UnloadableService(Topics topics, ServiceLoadException e) {
+        super(topics);
+        this.serviceLoadException = e;
+        this.exceptionTimestamp = Instant.now().toEpochMilli();
+        logger.atError().log(exceptionTimestamp);
+    }
+
+    @Override
+    protected void install() {
+        serviceErrored(serviceLoadException.getMessage());
+    }
+
+    @Override
+    public int bootstrap() {
+        return REQUEST_RESTART;
+    }
+
+    @Override
+    public boolean isBootstrapRequired(Map<String, Object> newServiceConfig) {
+        if (!PLUGIN.name().equals(newServiceConfig.get(SERVICE_TYPE_TOPIC_KEY))) {
+            logger.atInfo().log("Bootstrap is not required: component type is not Plugin");
+            return false;
+        }
+        if (!newServiceConfig.containsKey(VERSION_CONFIG_KEY)) {
+            logger.atInfo().log("Bootstrap is not required: config incomplete with no version");
+            return false;
+        }
+        String newVersion = newServiceConfig.get(VERSION_CONFIG_KEY).toString();
+        if (Utils.stringHasChanged(Coerce.toString(getConfig().find(VERSION_CONFIG_KEY)), newVersion)) {
+            logger.atInfo().log("Bootstrap is required: plugin version changed");
+            return true;
+        }
+        try {
+            Path pluginJar = config.getContext().get(NucleusPaths.class).artifactPath(new ComponentIdentifier(getName(),
+                    new Semver(newVersion))).resolve(getName() + JAR_FILE_EXTENSION);
+
+            if (!pluginJar.toFile().exists() || !pluginJar.toFile().isFile()) {
+                logger.atInfo().log("Bootstrap is not required: plugin JAR not found " + pluginJar);
+                return false;
+            }
+
+            if (Files.getLastModifiedTime(pluginJar).toMillis() > exceptionTimestamp) {
```

Comment:
Do we need to check if the jar is modified or is this covered by version check? If there's a different version of the plugin component, downloader would download the jar in the new version's artifact location even if the jar wasn't modified between versions. It will be very rare that the jar already existed on the device because the version had previously been in use and didn't get wiped out during component store cleanup.

Reply:
New version of plugin should be downloaded to a version-specific artifact location, and should be covered by version check.
This is for cases where there's no change in plugin version, but jar was corrupted previously somehow and to be fixed now in deployments.
The service config input here is the fully resolved config, which can be picked up from existing components as well as from the deployment document. This condition is to check if the deployment document contains fix to the broken plugin.